Fix 103 menu view nazaruk

Andrii35 opened the pull request:
Add drop-down
Fix view (background without image)
Don't need here DB migration 
Price validation, pagination, English gramms and images saved local are allready implemented.

Andrii35 committed:
change generationType to Identity

Andrii35 committed:
Merge branch 'dev' into fix_8_menu_view_nazaruk

Andrii35 committed:
fix test method

Andrii35 committed:
added changes to websecurityconfig

Andrii35 committed:
made chnages in entities by relationships

Andrii35 committed:
Merge branch 'dev' into fix_8_menu_view_nazaruk

Andrii35 committed:
fixed camelcase in categoryfoodcontroller

Andrii35 committed:
fixed menu view background

Andrii35 committed:
deleted bin order

Andrii35 committed:
fix camelCase in Entities

Andrii35 committed:
add some minor changes to seed file

Andrii35 committed:
add drop-down button

Andrii35 committed:
added posibility to choose dish on categoryFood

Andrii35 committed:
added dropdown to menu-header

Andrii35 committed:
fix merge conflict

Andrii35 committed:
change function name

Andrii35 committed:
button add in menu view made disabled

Andrii35 committed:
added db migration

Andrii35 committed:
fix minor issues for dropdown

Andrii35 committed:
add query to categoryfood controller and delete dep for DB migration

Andrii35 committed:
fix merge conflict in seed files

vitaliimak commented on server/src/main/java/com/lunchtime/controllers/MenuItemDishController.java:
```diff
@@ -20,7 +21,7 @@
 
 @RestController
-@RequestMapping("/api/menuitemdish")
+@RequestMapping("/api/menuItemDish")
```
Use lowercase in URL path

Andrii35 replied:
Fixed client and server part.

vitaliimak replied:
Use "Find in all project" for finding all paths

Andrii35 replied:
Thanks, fixed.

vitaliimak commented on server/src/main/java/com/lunchtime/models/Dish.java:
```diff
@@ -6,6 +6,7 @@
 import lombok.Getter;
 import lombok.Setter;
 import org.hibernate.annotations.ColumnDefault;
+import org.hibernate.annotations.OnDelete;
```
Remove redundant import

Andrii35 replied:
Fixed.

vitaliimak commented on server/src/main/java/com/lunchtime/models/Dish.java:
```diff
@@ -31,9 +32,10 @@
     public Dish() {
```
Could you place methods after fields?

Andrii35 replied:
Done.

vitaliimak commented on server/src/main/java/com/lunchtime/models/Dish.java:
```diff
@@ -31,9 +32,10 @@
     @NotNull
     @JoinColumn(name = "categoryfood_id", referencedColumnName = "id")
-    private CategoryFood categoryfood;
+
```
Is this blank line need?

Andrii35 replied:
Don't need. Removed

vitaliimak commented on server/src/main/java/com/lunchtime/models/MenuItemDish.java:
```diff
@@ -47,11 +40,11 @@
     public MenuItemDish() {
```
Could you place methods after fields?
https://google.github.io/styleguide/javaguide.html#s3.4.2-ordering-class-contents

Andrii35 replied:
Fixed

vitaliimak commented on server/src/main/java/com/lunchtime/seed/DatabaseSeed.java:
```diff
@@ -18,8 +18,7 @@
     String[] dishesName = new String[] {"Salami", "Margarita", "Manhattan", "Soup", "Hamburger", "Ice"};
-    String[] categoryFood = new String[] {"Pizza", "Pizza", "Pizza",
-                                          "Main course", "Snacks", "Dessert"};
+    String[] categoryFood = new String[] {"Pizza", "Pizza", "Pizza", "Main course", "Snacks", "Dessert"};
```
Why you add 3 "Pizza"? This file generates in fresh start, you should remove redundant categories

Andrii35 replied:
This code is on dev and in steging server. I have added here one "Pizza" to category but it works not  with DB created on dev. It don't shows any dishes in menu. I think it will be happen on staging server to.  Therefore, I marked the extra fields in the database as isDeleted "true". I have merged dev to my branch and it works as well.  
On staging server we will not create new DB. It will be added to existing DB, that 2 extra fields in categoryfood table are maked as isDeleted. Is that supposed to work that way?

vitaliimak replied:
This code invokes only when fresh project starts, it doesn't affect already created database. You should remove unnecessary items from code

Andrii35 replied:
I have deleted, it works, thank you.

vitaliimak commented on server/src/main/java/com/lunchtime/service/MenuItemDishService.java:
```diff
@@ -14,4 +14,7 @@
 
     Page<MenuItemDish> findByRestaurantId(Long id, Pageable pageable);
+
+    Page<MenuItemDish> findMenuItemDishesByRestaurantIdAndDish_CategoryFood_Name(
```
Follow Java style guide in method naming
https://google.github.io/styleguide/javaguide.html#s5.2.3-method-names

vitaliimak replied:
https://github.com/Social-projects-Rivne/Lunch_Time/commit/42f8a1365e8252f91ce9643fb8d14a67d6f29d37#diff-045fb6a4100c4cbfc55fe645831ca249R8

Andrii35 replied:
Thanks. Have forgoten about. Fixed

vitaliimak reviewed (changes requested):
I've created DB in dev branch, register new user, switch to your branch and got that errors
![image](https://user-images.githubusercontent.com/6152066/80181530-b3f38800-860d-11ea-8e80-5676b20d802e.png)

Could you remove redundant blank lines and unused imports in all your code?

Andrii35 committed:
return lower case in URL path

Andrii35 committed:
remove redundant import

Andrii35 committed:
remove methods after fields

Andrii35 committed:
change name by following of java code convention

Andrii35 committed:
changed key from name to id

Andrii35 committed:
fix minor issues with button

Andrii35 committed:
changes in seed categoryfood

vitaliimak commented:
```
12:48 % mvn clean spring-boot:run
[INFO] Scanning for projects...
[INFO]
[INFO] ----------------------< com.LunchTime:LunchTime >-----------------------
[INFO] Building LunchTime 0.0.1-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO]
[INFO] --- maven-clean-plugin:3.1.0:clean (default-clean) @ LunchTime ---
[INFO] Deleting /home/vitalii/Projects/study/Lunch_Time/server/target
[INFO]
[INFO] >>> spring-boot-maven-plugin:2.2.4.RELEASE:run (default-cli) > test-compile @ LunchTime >>>
[INFO]
[INFO] --- maven-checkstyle-plugin:3.1.1:check (validate) @ LunchTime ---
[INFO] Starting audit...
[ERROR] /home/vitalii/Projects/study/Lunch_Time/server/src/main/java/com/lunchtime/seed/DatabaseSeed.java:127:13: WhitespaceAround: 'if' is not followed by whitespace. Empty blocks may only be represented as {} when not part of a multi-block statement (4.1.3) [WhitespaceAround]
[ERROR] /home/vitalii/Projects/study/Lunch_Time/server/src/main/java/com/lunchtime/seed/DatabaseSeed.java:127:17: WhitespaceAround: '==' is not followed by whitespace. Empty blocks may only be represented as {} when not part of a multi-block statement (4.1.3) [WhitespaceAround]
[ERROR] /home/vitalii/Projects/study/Lunch_Time/server/src/main/java/com/lunchtime/seed/DatabaseSeed.java:127:17: WhitespaceAround: '==' is not preceded with whitespace. [WhitespaceAround]
[ERROR] /home/vitalii/Projects/study/Lunch_Time/server/src/main/java/com/lunchtime/seed/DatabaseSeed.java:127:22: WhitespaceAround: '||' is not followed by whitespace. Empty blocks may only be represented as {} when not part of a multi-block statement (4.1.3) [WhitespaceAround]
[ERROR] /home/vitalii/Projects/study/Lunch_Time/server/src/main/java/com/lunchtime/seed/DatabaseSeed.java:127:25: WhitespaceAround: '==' is not followed by whitespace. Empty blocks may only be represented as {} when not part of a multi-block statement (4.1.3) [WhitespaceAround]
[ERROR] /home/vitalii/Projects/study/Lunch_Time/server/src/main/java/com/lunchtime/seed/DatabaseSeed.java:127:25: WhitespaceAround: '==' is not preceded with whitespace. [WhitespaceAround]
Audit done.
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  19.828 s
[INFO] Finished at: 2020-04-24T12:48:56+03:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.1:check (validate) on project LunchTime: Failed during checkstyle execution: There are 6 errors reported by Checkstyle 8.29 with checkstyle.xml ruleset. -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
```

Andrii35 committed:
fixed minor issues with checkstyle

Andrii35 commented:
Fixed checkstyle errors.

Andrii35 committed:
Merge branch 'dev' into fix_103_menu_view_nazaruk

vitaliimak reviewed (changes requested):
Could you merge current dev branch into your feature branch?

Andrii35 committed:
Merge branch 'dev' into fix_103_menu_view_nazaruk

Andrii35 committed:
change query for drop down after merge

Andrii35 committed:
minot changes to make button disabled in menu-item-dish

Andrii35 committed:
add migration for updating of status of category_food

VozNick commented on server/src/main/java/com/lunchtime/service/MenuItemDishService.java:
```diff
@@ -14,4 +14,7 @@
 
     Page<MenuItemDish> findByRestaurantId(Long id, Pageable pageable);
+
+    Page<MenuItemDish> findDishesByRestaurantIdAndDishCategoryName(String name,
```
I think 3rd method is quite long...
dishesByRestaurantId is redundant because it's clear that we are searching dishes in specific rest.
maybe "findByCategory" or "findByRestCategory"

Andrii35 replied:
The name may be long, but other developers immediately understand what the method does. I suggest shortening the name to "findDishesByRestaurantAndCategory"

vitaliimak replied:
> dishesByRestaurantId is redundant because it's clear that we are searching dishes in specific rest.

It isn't clear. There is parameter `id` in this method, it isn't explain what it is. 
`id` - `restaurantId` or `dishId`, or `categoryId`?

Andrii35 replied:
Thank you both, I have changed the name to "findDishesByRestaurantIdAndCategoryName". I will push changes in few minutes.

vitaliimak commented on server/src/main/java/com/lunchtime/controllers/MenuItemDishController.java:
```diff
@@ -56,12 +55,27 @@ public MenuItemDishController(MenuItemDishService menuItemDishService) {
         MenuItemDish newMenuItemDish = menuItemDishService.save(menuItemDish);
         return  ResponseEntity
-               .created(new URI("/api/menuitemdish"))
+               .created(new URI("/api/menuItemDish"))
```
URL should be lowercase letters

Andrii35 replied:
Fixed.

vitaliimak commented on server/src/main/java/com/lunchtime/models/Restaurant.java:
```diff
@@ -1,6 +1,7 @@
 package com.lunchtime.models;
 
 import java.time.Instant;
+import java.util.Set;
```
You can use Analyze -> Inspect code in IDEA to find all unused imports and remove your redundant code

Andrii35 replied:
Made it.

vitaliimak commented on server/src/test/java/com/lunchtime/controllers/CategoryFoodControllerTest.java:
```diff
@@ -21,10 +23,14 @@
     public void setUp() {
         initMocks(this);
         categoryFoodControllerUnderTest = new CategoryFoodController(mockCategoryFoodService);
+        mockCategoryFoodController = new CategoryFoodController(mockCategoryFoodService);
```
Why you add this mock, but don't use it in tests?

Andrii35 replied:
Fixed it.

Andrii35 committed:
deleted unused params in seed file  and fixed minor issues

vitaliimak commented on client/src/components/menu-views/menu-item-dish.js:
```diff
@@ -35,18 +35,21 @@ class MenuItemDish extends Component {
-                <Button variant="primary">
+                <Button
+                  variant="primary"
+                  disabled
```
Why this button always disabled?

Andrii35 replied:
Made button aktiv for loged users.

vitaliimak reviewed (changes requested):
You have issue - валідація на ціну (формат гроші), but you still use `@Column(name = "portion_price", length = 20)
    private Long portionPrice;`
Is your dish price can't have coins?

обновити чекбокси для меню (додати можливість вибору к-сті страв) - it doesn't work

Andrii35 committed:
delete redundant lields

Andrii35 committed:
Button add in menu view maked aktiv for loged users

Andrii35 commented:
> You have issue - валідація на ціну (формат гроші), but you still use `@Column(name = "portion_price", length = 20) private Long portionPrice;`
> Is your dish price can't have coins?
> 
> обновити чекбокси для меню (додати можливість вибору к-сті страв) - it doesn't work

By price validation i have deleted length. I use here anotation @DecimalMin, that owner can not add negative price.
I have changed checkboxes to Button "add".  I think, that add dish to menu order should be realised in the task #137. As an authorized user of the system I want to have a possibility to add dishes to order.

vitaliimak commented:
@Andrii35 
> I have changed checkboxes to Button "add". I think, that add dish to menu order should be realised in the task #137. As an authorized user of the system I want to have a possibility to add dishes to order.

Ok, I understand. You should update your task, add this notes to comments

> By price validation i have deleted length. I use here anotation @DecimalMin, that owner can not add negative price.

Is dish can have price 20,50grn?

Andrii35 committed:
changed price to BigDecimal

Andrii35 commented:
> @Andrii35
> 
> > I have changed checkboxes to Button "add". I think, that add dish to menu order should be realised in the task #137. As an authorized user of the system I want to have a possibility to add dishes to order.
> 
> Ok, I understand. You should update your task, add this notes to comments
> 
> > By price validation i have deleted length. I use here anotation @DecimalMin, that owner can not add negative price.
> 
> Is dish can have price 20,50grn?

Dish can't have price 20,50. I change Long to  BigDecimal with help of migration, it will be recomended to use with money.

VozNick commented:
I created user (logged in). Add button is active, but when I click on it - nothing happens

Andrii35 commented:
> I created user (logged in). Add button is active, but when I click on it - nothing happens

It should not something happen. In this fix to task menu view button changed from disabled for unloged user to activ for loged user. I have changed checkboxes to Button "add". I think, that add dish to menu order should be realised in the task #137. As an authorized user of the system I want to have a possibility to add dishes to order.

Andrii35 committed:
Merge branch 'dev' into fix_103_menu_view_nazaruk

vitaliimak commented on server/src/main/resources/db/migration/V3__update_categoryfood.sql:
```diff
@@ -0,0 +1,3 @@
+update category_food
+set is_deleted = true
+where id = 2 or id = 3;
```
Since you remove redundant categories from seed file, this code isn't necessary. If new database will be created in new server, others categories will have id = 2 and 3
For updating production database you should use DELETE method in CategoryController

Andrii35 replied:
I disagree with you. We are currently working on an existing database. You suggested that we make all changes to the database through migration. If you create a new server or new DB each time, I will delete the existing data through the controller.

vitaliimak replied:
In migration, we should change database scheme not database content

Andrii35 replied:
Ok, added DELETE to CategoryFoodController. Updating production database will be done after PR via DELETE method.

vitaliimak commented on server/src/main/java/com/lunchtime/repository/CategoryFoodRepository.java:
```diff
@@ -1,12 +1,16 @@
+import org.springframework.data.jpa.repository.Query;
 import org.springframework.stereotype.Repository;
 
 import java.util.List;
```
Is this import need?

Andrii35 replied:
Yes, i use here  @Query("select c from CategoryFood c where c.isDeleted = false ")

Andrii35 replied:
Sory, List don't need here

vitaliimak commented on server/src/main/java/com/lunchtime/models/MenuItemDish.java:
```diff
@@ -28,11 +22,12 @@
     @Column(name = "portion_size", length = 30)
     private String portionSize;
 
+
```
Remove redundant lines

vitaliimak replied:
I see few redundant lines in this file

vitaliimak commented on server/src/main/java/com/lunchtime/models/MenuItemDish.java:
```diff
@@ -28,11 +22,12 @@
-    @Column(name = "portion_price", length = 20)
-    private Long portionPrice;
+    @Column(name = "portion_price")
+    private BigDecimal portionPrice;
```
Why BigDecimal isn't good decision - 10 UAH 2322 cents
![image](https://user-images.githubusercontent.com/6152066/80457132-7bb6b700-8937-11ea-9679-56b588667fed.png)

I'm not sure, but maybe class Currency can help to you https://docs.oracle.com/javase/6/docs/api/java/util/Currency.html

Or you can create special class Currency and validate this field

Andrii35 replied:
I have changed BigDecimal to Double. Added field currency and validate it with @Currency

vitaliimak commented on server/src/main/java/com/lunchtime/controllers/MenuItemDishController.java:
```diff
@@ -56,6 +54,21 @@ public MenuItemDishController(MenuItemDishService menuItemDishService) {
 
+    @GetMapping("/category")
+    public ResponseEntity<Page<MenuItemDish>> getAllByDishCategoryName(
+        @RequestParam("name") String name,
```
Maybe it isn't important, but your API is case sensitive (it returns "Pizza", but doesn't return "pizza")

Andrii35 replied:
I think we don't need it here .  On client part user choose between existing categories and don't type it manually.

vitaliimak commented on bin/deploy.sh:
```diff
@@ -1,21 +0,0 @@
-#! /bin/bash
```
Why you remove this file?

Andrii35 replied:
I thot, that bin order not necessary here. Are we needed it here?

vitaliimak replied:
You don't want to deploy application to staging server? Why you change your mind?)
https://github.com/Social-projects-Rivne/Lunch_Time/pull/45

Andrii35 replied:
Ok, it will be returned

vitaliimak reviewed (changes requested):
You need to validate price. Your PR breaks deployment process

Andrii35 committed:
Merge branch 'dev' into fix_103_menu_view_nazaruk

Andrii35 committed:
add DELETE to categoryFoodController

Andrii35 committed:
made minor changes to show currency on clientpart

Andrii35 committed:
return bin folder

Andrii35 committed:
change portion type to double and add currency

Andrii35 committed:
merge dev

Andrii35 committed:
fix minor issues in seed and test

Andrii35 committed:
fix minor changes with types by DB migration

Andrii35 committed:
deleted annotation for currency

Andrii35 committed:
fix merge conflicts

vitaliimak commented on server/src/main/java/com/lunchtime/models/MenuItemDish.java:
```diff
@@ -1,21 +1,15 @@
 import lombok.Getter;
 import lombok.Setter;
 import org.hibernate.annotations.ColumnDefault;
+import org.hibernate.validator.constraints.Currency;
```
Unused import :(

Andrii35 replied:
I will fix it in the next fix task #115. It is on hold now. I have added allready in trello comment.

vitaliimak reviewed (approved):
Cool! Validation works too
![image](https://user-images.githubusercontent.com/6152066/80627868-37233c80-8a59-11ea-9f7d-9736390c39e5.png)
